pybossa/sched.py

```diff
@@ -65,11 +64,25 @@ def new_task(project_id, sched, user_id=None, user_ip=None,
     }
     scheduler = sched_map.get(sched, sched_map['default'])
 
-    # This is here for testing. It removes the random variable to make testing deterministic.
     project = project_repo.get(project_id)
+    user = user_repo.get(user_id)
```

Comment:
is there any reason why this was moved here?

Reply:
This needs to happen whenever a new task is requested. If it is outside this function then every caller of this function will have to do it.

Reply:
`/api/newtask` is the only way to request a task for contribution

Reply:
I'm seeing this code do it too:

https://github.com/bloomberg/pybossa/blob/78a328958ae45fd0860d435a1fddb3e3f03e2502/pybossa/view/projects.py#L1047-L1059

That's why I needed to move that code into `sched.new_task()`

Reply:
This code is certainly relevant. Here the `new_task` function is used to check whether the user has tasks left - the problem is that now the function locks the task as for some schedulers, and this does not work well with new gold task features. I don't think adding more side effects to this function is the solution though.
Ideally the locking should be done outside of this function, but that's a bigger change. Perhaps we can change the call to `new_task` here with a function that has no such side effects, like [`n_available_tasks`](https://github.com/bloomberg/pybossa/blob/master/pybossa/cache/helpers.py#L32).
Or we remove this call since task presenters are supposed to handle this case anyway.

Reply:
I deleted the function call and just made it `False`.